pandas/_libs/tslibs/conversion.pyx

```diff
@@ -642,3 +644,94 @@ cpdef inline datetime localize_pydatetime(datetime dt, tzinfo tz):
     elif isinstance(dt, ABCTimestamp):
         return dt.tz_localize(tz)
     return _localize_pydatetime(dt, tz)
+
+
+cdef tzinfo validate_tzout(
+        tzinfo tz_in,
+        tzinfo tz_out,
+        bint found_naive,
+        bint found_tz,
+        bint utc_convert,
+):
+    """
+    Validate that ``tz_in`` can be converted/localized to ``tz_out``.
+
+    Parameters
+    ----------
+    tz_in : tzinfo
+        Timezone info of element being processed.
+    tz_out : tzinfo
+        Timezone info of output.
+    found_naive : bool
+        Whether a timezone-naive element has been found so far.
+    found_tz : bool
+        Whether a timezone-aware element has been found so far.
+    utc_convert : bool
+        Whether to convert/localize to UTC.
+
+    Returns
+    -------
+    tz_info
+        Timezone info of output.
+
+    Raises
+    ------
+    ValueError
+        If ``tz_in`` can't be converted/localized to ``tz_out``.
+    """
+    if tz_in is not None:
+        if utc_convert:
+            pass
+        elif found_naive:
+            raise ValueError('Tz-aware datetime.datetime '
+                             'cannot be converted to '
+                             'datetime64 unless utc=True')
+        elif tz_out is not None and not tz_compare(tz_out, tz_in):
+            raise ValueError('Tz-aware datetime.datetime '
+                             'cannot be converted to '
+                             'datetime64 unless utc=True')
+        else:
+            tz_out = tz_in
+    else:
+        if found_tz and not utc_convert:
+            raise ValueError('Cannot mix tz-aware with '
+                             'tz-naive values')
+    return tz_out
+
+cdef int64_t parse_pydatetime(
+        object val,
+        npy_datetimestruct *dts,
+        bint utc_convert,
+) except *:
```

Comment:
As an aside I wonder how much overhead we have from using `except *` instead of explicitly checking for a return sentinel like -1 before going down that route

Reply:
looks like `except? -1` works, thanks! the `?` is needed because
```
'1969-12-31 23:59:59.999999999'
```
would return `-1` (and it would be valid!)

Reply:
Hmm OK. I am not familiar with all the functions you are calling underneath this but the difference is that `except *` loosely gets translated into `if (PyErr_Occurred() { // handler }` whereas `except -1` would be `if ((result == -1) && (PyErr_Occurred()) { // handler }`

The former will be a cheaper check since it should short circuit most of the time, and also because the equality evaluation will be much cheaper than PyErr_Occurred (which also needs the GIL). However, it also requires that your arguments actually do return -1 in case of an error and not some other sentinel

Reply:
On deeper review I think this works because `ts.ensure_reso` calls `convert_reso` from the np_datetime module's `convert_reso`, which is also a `except? -1`. That's a bit confusing from a code structure perspective - would be better if `ts.convert_reso` wasn't void and we could inspect each step along the way...but that's for another PR

Reply:
It looks to me like this is working, because if I run
```python
to_datetime([datetime(9999, 1, 1)])
```
then we do indeed get to

https://github.com/pandas-dev/pandas/blob/51db700e2d53c75e78058d98be9ec95cb144d87d/pandas/_libs/tslib.pyx#L666-L683

and the exception is raised, whereas
```python
In [2]: to_datetime([Timestamp('1969-12-31 23:59:59.999999999')])
Out[2]: DatetimeIndex(['1969-12-31 23:59:59.999999999'], dtype='datetime64[ns]', freq=None)
```
parses fine

Tests that cover these inputs are:

https://github.com/pandas-dev/pandas/blob/371a8f2e052bd3e02623280b848f4912de279f86/pandas/tests/indexes/period/methods/test_to_timestamp.py#L86-L95

https://github.com/pandas-dev/pandas/blob/a37b78d534cd25e21d2ab64318ceae808df46c89/pandas/tests/tools/test_to_datetime.py#L136-L146